### PR TITLE
Automatic image upload from rich text editor

### PR DESCRIPTION
**Pull request by szabozoltan69**
Refers to https://github.com/IFRCGo/go-api/issues/1804

### CODE REVIEW

**Comment by batpad** [comment]
@thenav56 could you take a look?

I am guessing we probably want to upload the images to Azure Blob Storage or so, rather than just on the local file-system? Otherwise, broadly this looks good and like what we need to do here.

File: api/views.py [hunk 1]
```diff
@@ -50,6 +53,45 @@ def unauthorized(message="You must be logged in"):
...
+
+        with open(file_path, 'wb+') as f:
+            for chunk in file_obj.chunks():
+                f.write(chunk)
+
+        return JsonResponse({
+            'message': 'Image uploaded successfully',
+            'location': file_url
+        })
+    return JsonResponse({'detail': "Wrong request"})
```
Comment (thenav56) [inline api/views.py:92]: Hi @szabozoltan69

For uploading, we can directly use the storage class instead.
Something like this
```python3
from django.core.files.storage import FileSystemStorage, get_storage_class

DefaultMediaStorageClass = get_storage_class()


def get_full_media_url(media_path):
    if DefaultMediaStorageClass == FileSystemStorage:
        return f"{settings.HTTP_PROTOCOL}://{settings.API_FQDN}{media_path}"
    # For other storage. Assuming they provide absolute URL
    return media_path


def func...():
    ...
    storage = DefaultMediaStorageClass()
    storage.save(file_path, content_file)
    file_url = get_full_media_url(file_path)
```

> WHERE `get_storage_class` provides storage defined using `DEFAULT_FILE_STORAGE`
https://docs.djangoproject.com/en/3.2/ref/files/storage/#django.core.files.storage.get_storage_class

File: main/settings.py [hunk 2]
```diff
@@ -282,6 +282,7 @@
 ]
 
 TINYMCE_DEFAULT_CONFIG = {
+    'images_upload_url': '/upload_image',  # Image upload address in Django route
```
Comment (thenav56) [inline main/settings.py:285]: Maybe we can use reverse_lazy.
https://docs.djangoproject.com/en/3.2/ref/urlresolvers/#reverse-lazy

File: api/views.py [hunk 3]
```diff
@@ -50,6 +53,45 @@ def unauthorized(message="You must be logged in"):
     return JsonResponse({"statusCode": 401, "error_message": message}, status=401)
 
 
+@csrf_exempt
+def upload_image(request):
```
Comment (thenav56) [inline api/views.py:57]: - Restrict with logged-in users only
- Track these uploads (Suggestion)
  - Maybe create a table with a columns file path, created by, created at

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: api/views.py (hunk 3 above)
Comment (thenav56) [reply api/views.py:57]: Maybe we can use images_upload_handler to provide CSRF+Auth
https://www.tiny.cloud/docs/demo/local-upload/#

File: api/views.py (hunk 1 above)
Comment (szabozoltan69) [reply api/views.py:92]: Thanks @thenav56

I wonder how content_file variable should have it's content.

Comment (thenav56) [reply api/views.py:92]: I get confused with this, but something like this should do :D 
```python
from django.core.files.base import ContentFile

ContentFile(
        file.read().encode('utf-8')
    )

```

Comment (szabozoltan69) [reply api/views.py:92]: Well, I could not use this file.read (not obvious which would be the "file"), but maybe file_obj.chunks() could be used like:
```
        with open(file_url, 'wb+') as f:
            for chunk in file_obj.chunks():
                f.write(chunk)
```

Comment (thenav56) [reply api/views.py:92]: Maybe `ContentFile(file_obj)` works?
Or directly `storage.save(file_path, file_obj)`

File: main/settings.py (hunk 2 above)
Comment (szabozoltan69) [reply main/settings.py]: I was not able to get this reverse_lazy work. Please give me some more hints.
I gave a name in urls.py, but there are other tasks also.
```
    url(r"^upload_image", upload_image, name="upload_image"),
```

**Comment by szabozoltan69** [comment]
The "rest_framework.exceptions.PermissionDenied: CSRF Failed: CSRF token missing or incorrect" should be fixed.

File: main/settings.py (hunk 2 above)
Comment (thenav56) [reply main/settings.py]: Hey @szabozoltan69 Did you mean there are other URLs with the same name? Maybe we can use a different name instead? Like "tinymc_upload_image"
Let me know if I understood incorrectly.

Comment (szabozoltan69) [reply main/settings.py]: Hey @thenav56 , there is no other same name, just for the sake of reverse_lazy I gave a name to the endpoint. But I could not do the next steps about it. Maybe not even meeded to be named.

Comment (szabozoltan69) [reply main/settings.py]: As "other tasks" I meant other developing tasks for sake of reverse_lazy.

Comment (thenav56) [reply main/settings.py]: Okay. Let's leave with a TODO there then.

**Comment by szabozoltan69** [comment]
@batpad @thenav56 Please have a look at this, mainly the CSRF "protection" is questionable.

**Comment by tnagorra** [comment]
@szabozoltan69 Aren't we disabling "image upload from rich text editors" for the time being?

cc: @batpad @frozenhelium @thenav56

**Comment by szabozoltan69** [comment]
Yes, we disabled "image paste from rich text editors". The image link placing is still ok from Admin page (and I think it should be also from the client side), but let's close this ticket due to the "automation" is not on our plate.

**Pull request closed without merging**